In [25]:
import pandas as pd       
import numpy as np        
import matplotlib.pyplot as plt  
import seaborn as sns         
import warnings
warnings.filterwarnings('ignore')


In [26]:
import os
print(os.getcwd())

C:\vw_project\notebooks


In [27]:
import os
os.chdir(r'C:\vw_project\notebooks')
print("Working directory set to:", os.getcwd())

Working directory set to: C:\vw_project\notebooks


In [28]:
df = pd.read_csv(r'C:\vw_project\data\raw\Volkswagen_Stock.csv')

In [29]:
df = pd.read_csv('../data/raw/Volkswagen_Stock.csv')

In [30]:
print('Shape:', df.shape)
print('Columns:', df.columns.tolist())
df.head() 


Shape: (7930, 25)
Columns: ['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Daily_Return_Pct', 'Daily_Range', 'MA_20', 'MA_50', 'MA_200', 'EMA_12', 'EMA_26', 'MACD', 'MACD_Signal', 'BB_Upper', 'BB_Lower', 'BB_Width', 'RSI_14', 'Volume_MA_20', 'Cumulative_Return_Pct', 'Year', 'Month', 'Quarter', 'Day_of_Week']


,Date,Open,High,Low,Close,Volume,Daily_Return_Pct,Daily_Range,MA_20,MA_50,...,BB_Upper,BB_Lower,BB_Width,RSI_14,Volume_MA_20,Cumulative_Return_Pct,Year,Month,Quarter,Day_of_Week
0,1995-03-24,17.1538,17.2050,17.0000,17.4095,0,NaN,0.2050,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,0.00,1995,3,1,Friday
1,1995-03-27,17.6907,17.8952,17.5884,17.2050,0,-1.1746,0.3068,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,-1.17,1995,3,1,Monday
2,1995-03-28,17.5884,17.6038,17.2050,17.2050,0,0.0000,0.3988,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,-1.17,1995,3,1,Tuesday
3,1995-03-29,17.2050,17.2050,17.2050,17.2050,0,0.0000,0.0000,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,-1.17,1995,3,1,Wednesday
4,1995-03-30,17.1794,17.7930,17.1794,17.7418,10,3.1200,0.6136,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,1.91,1995,3,1,Thursday


In [31]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7930 entries, 0 to 7929
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Date                   7930 non-null   str    
 1   Open                   7930 non-null   float64
 2   High                   7930 non-null   float64
 3   Low                    7930 non-null   float64
 4   Close                  7930 non-null   float64
 5   Volume                 7930 non-null   int64  
 6   Daily_Return_Pct       7929 non-null   float64
 7   Daily_Range            7930 non-null   float64
 8   MA_20                  7911 non-null   float64
 9   MA_50                  7881 non-null   float64
 10  MA_200                 7731 non-null   float64
 11  EMA_12                 7930 non-null   float64
 12  EMA_26                 7930 non-null   float64
 13  MACD                   7930 non-null   float64
 14  MACD_Signal            7930 non-null   float64
 15  BB_Upper       

In [32]:
df.describe()

,Open,High,Low,Close,Volume,Daily_Return_Pct,Daily_Range,MA_20,MA_50,MA_200,...,MACD_Signal,BB_Upper,BB_Lower,BB_Width,RSI_14,Volume_MA_20,Cumulative_Return_Pct,Year,Month,Quarter
count,7930.000000,7930.000000,7930.000000,7930.000000,7.930000e+03,7929.000000,7930.000000,7911.000000,7881.000000,7731.000000,...,7930.000000,7911.000000,7911.000000,7911.000000,7842.000000,7.911000e+03,7930.00000,7930.000000,7930.000000,7930.000000
mean,112.856102,114.505508,111.134678,112.797557,9.671652e+05,0.055331,3.370831,112.924272,113.124163,114.163468,...,0.075636,121.797600,104.050944,17.746654,51.597576,9.694319e+05,547.90810,2010.096847,6.515132,2.507440
std,67.426068,69.107623,65.953704,67.418700,2.467386e+06,2.809585,6.321663,66.964782,66.462505,64.044665,...,3.884136,75.502308,60.192527,26.652385,18.251381,1.363345e+06,387.25232,8.951317,3.431853,1.114351
min,17.153800,17.205000,17.000000,17.205000,0.000000e+00,-22.659300,0.000000,17.644700,17.702900,20.845000,...,-26.370000,17.741800,14.366000,0.000000,1.597400,0.000000e+00,-1.17000,1995.000000,1.000000,1.000000
25%,53.150000,53.870000,52.282500,53.200000,6.808475e+04,-1.030700,1.290000,53.060000,53.816800,52.483450,...,-1.080600,57.013600,49.123150,6.688800,37.873475,7.917883e+04,205.58000,2002.000000,4.000000,2.000000
50%,110.325000,112.130000,108.700000,110.100000,1.902820e+05,0.000000,2.300000,111.343000,110.836000,113.816300,...,0.164450,118.145800,100.534200,12.056800,51.327650,2.259175e+05,532.41000,2010.000000,7.000000,3.000000
75%,155.400000,157.287500,153.915000,155.600000,1.455133e+06,1.110500,3.800000,155.228750,155.203200,157.448250,...,1.428650,166.111450,146.007100,20.056850,64.918450,1.730491e+06,793.76000,2018.000000,9.000000,3.000000
max,520.000000,635.000000,491.000000,520.000000,1.672419e+08,146.620800,310.010000,415.961500,347.293200,281.040000,...,40.213500,580.770800,304.171500,413.752200,100.000000,1.061406e+07,2886.87000,2026.000000,12.000000,4.000000


In [33]:
df.isnull().sum()

Date                       0
Open                       0
High                       0
Low                        0
Close                      0
Volume                     0
Daily_Return_Pct           1
Daily_Range                0
MA_20                     19
MA_50                     49
MA_200                   199
EMA_12                     0
EMA_26                     0
MACD                       0
MACD_Signal                0
BB_Upper                  19
BB_Lower                  19
BB_Width                  19
RSI_14                    88
Volume_MA_20              19
Cumulative_Return_Pct      0
Year                       0
Month                      0
Quarter                    0
Day_of_Week                0
dtype: int64

In [34]:
df['Date'] = pd.to_datetime(df['Date'])

In [35]:
df = df.set_index('Date')

In [36]:
df = df.sort_index()

In [37]:
print('Duplicates before:', df.index.duplicated().sum())
df = df[~df.index.duplicated(keep='first')]
print('Duplicates after:', df.index.duplicated().sum())


Duplicates before: 0
Duplicates after: 0


In [38]:
print('Missing Close values:', df['Close'].isnull().sum())
df['Close'] = df['Close'].ffill()


Missing Close values: 0


In [39]:
print(df.columns.tolist())

['Open', 'High', 'Low', 'Close', 'Volume', 'Daily_Return_Pct', 'Daily_Range', 'MA_20', 'MA_50', 'MA_200', 'EMA_12', 'EMA_26', 'MACD', 'MACD_Signal', 'BB_Upper', 'BB_Lower', 'BB_Width', 'RSI_14', 'Volume_MA_20', 'Cumulative_Return_Pct', 'Year', 'Month', 'Quarter', 'Day_of_Week']


In [40]:
invalid_ohlc = df[df['High'] < df['Low']]
print('Invalid OHLC rows:', len(invalid_ohlc))


Invalid OHLC rows: 0


In [41]:
zero_vol = df[df['Volume'] == 0]
print('Zero volume days:', len(zero_vol))


Zero volume days: 243


Data Quality Note — Volume Zeros
243 trading days recorded zero volume. For a major exchange-listed stock like Volkswagen, true zero-volume days are implausible. These are treated as missing data and replaced using forward fill (previous trading day's volume). This affects approximately 3% of records and is documented here for transparency.

In [42]:
print(zero_vol.index.min())   # earliest zero-volume date
print(zero_vol.index.max())   # latest zero-volume date
print(zero_vol.index.year.value_counts().sort_index())  # how many per year

1995-03-24 00:00:00
2025-12-18 00:00:00
Date
1995    109
1996     20
1997     12
1998     10
1999      8
2000      6
2001      8
2002      8
2003      8
2004      5
2005      3
2014      1
2015      1
2017      3
2018      1
2021      2
2022     10
2023      2
2024      8
2025     18
Name: count, dtype: int64


Data Quality Decision — Start Date
The 1995–1999 period contains severe volume data gaps (109 zero-volume days in 1995 alone, representing ~44% of that year's trading days). This indicates poor data digitisation from the early period rather than genuine market activity. Analysis has been restricted to 2000 onwards where data quality is reliable. This decision is conservative and prioritises analytical integrity over dataset size.

In [43]:
df = df[df.index.year >= 2000]

print(f'Dataset now runs from {df.index.min().date()} to {df.index.max().date()}')
print(f'Total trading days: {len(df)}')

Dataset now runs from 2000-01-03 to 2026-02-26
Total trading days: 6684


In [44]:
import numpy as np

print('Average volume before fix:', df['Volume'].mean().round(0))

df['Volume'] = df['Volume'].replace(0, np.nan)
df['Volume'] = df['Volume'].ffill()

print('Zero volume days after fix:', (df['Volume'] == 0).sum())
print('Average volume after fix:', df['Volume'].mean().round(0))

Average volume before fix: 913382.0
Zero volume days after fix: 0
Average volume after fix: 921909.0


In [45]:
print('Final missing values:')
print(df.isnull().sum())
print('Date range:', df.index.min(), 'to', df.index.max())
print('Total trading days:', len(df))


Final missing values:
Open                     0
High                     0
Low                      0
Close                    0
Volume                   0
Daily_Return_Pct         0
Daily_Range              0
MA_20                    0
MA_50                    0
MA_200                   0
EMA_12                   0
EMA_26                   0
MACD                     0
MACD_Signal              0
BB_Upper                 0
BB_Lower                 0
BB_Width                 0
RSI_14                   0
Volume_MA_20             0
Cumulative_Return_Pct    0
Year                     0
Month                    0
Quarter                  0
Day_of_Week              0
dtype: int64
Date range: 2000-01-03 00:00:00 to 2026-02-26 00:00:00
Total trading days: 6684


In [46]:
df.to_csv(r'C:\vw_project\data\processed\vw_stock_cleaned.csv')
print('Cleaned file saved successfully.')

Cleaned file saved successfully.


In [47]:
df['Volatility_30d'] = df['Daily_Return_Pct'].rolling(window=30).std()
print('Volatility column added.')

Volatility column added.


## Created Volatility_30d

This column was missing from the dataset. It calculates the standard deviation 
of daily returns over a rolling 30-day window, measuring how wildly the stock 
price is swinging at any given point in time.

High value = large unpredictable moves = high risk period (e.g. 2008, COVID-19)  
Low value  = calm steady movement = stable market conditions

30 days because it is responsive enough to catch real events quickly, 
but long enough to smooth out single-day noise.

In [48]:
df.to_csv(r'C:\vw_project\data\processed\vw_stock_features.csv')
print('File saved.')

File saved.
